[mlir] Expose type and attribute names in the MLIRContext and abstract type/attr classes

math-fehr opened the pull request:
This patch expose the type and attribute names in C++ as methods in the `AbstractType` and `AbstractAttribute` classes, and keep a map of names to `AbstractType` and `AbstractAttribute` in the `MLIRContext`. Type and attribute names should be unique.

It adds support in ODS to generate the `getName` methods in `AbstractType` and `AbstractAttribute`, through the use of two new variables, `typeName` and `attrName`. It also adds names to C++-defined type and attributes.

math-fehr committed:
Add names to AbtsractType and Type

math-fehr committed:
Add names to attributes too

math-fehr committed:
Fix unwnated change

math-fehr committed:
Cleanup AttrOrTypeDef

math-fehr committed:
Add comment

math-fehr committed:
Insert attribute name to context

math-fehr commented:
RFC here: https://discourse.llvm.org/t/rfc-exposing-type-and-attribute-names-in-c/74896

math-fehr committed:
Add missing name in toy tutorial type

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTI.h:
```diff
@@ -55,6 +55,10 @@ class DataLayoutEntryAttr
   /// Prints this attribute.
   void print(AsmPrinter &os) const;
+
+  static constexpr StringRef getAttrName() {
```
Why not just `getName()` so this is the same API on Types and Attribute (which is always more friendly to generic code).

joker-eph replied:
Reasonnable I guess

joker-eph commented on mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp:
```diff
@@ -131,6 +138,8 @@ struct TypeNoLayout : public Type::TypeBase<TypeNoLayout, Type, TypeStorage> {
   using Base::Base;
 
+  static constexpr StringRef getTypeName() { return "test.no_layout"; }
+
```
One thing with these accessors is that they will interfere with the auto-generated accessors for the parameters, what happens in this case?
Should ODS forbid a parameter named `name`?

makslevental commented on mlir/include/mlir/IR/BuiltinAttributes.td:
```diff
@@ -64,6 +64,7 @@ def Builtin_AffineMapAttr : Builtin_Attr<"AffineMap", [
     AffineMap getAffineMap() const { return getValue(); }
   }];
   let skipDefaultBuilders = 1;
+  let attrName = "builtin.affine_map";
```
So I'm just curious - why can't the same reparameterization that's done for `Builtin_Type` (taking the mnemonic as an arg) be done for `Builtin_Attr`?

math-fehr replied:
Oh yes, that's an overlook on my side, I fixed it now to work like `Builtin_Type`!

kuhar commented on mlir/examples/toy/Ch7/include/toy/Dialect.h:
```diff
@@ -72,6 +72,9 @@ class StructType : public mlir::Type::TypeBase<StructType, mlir::Type,
   size_t getNumElementTypes() { return getElementTypes().size(); }
+
+  /// Returns the name of this struct type.
+  static constexpr StringRef getTypeName() { return "toy.struct"; }
```
Why not `StringLiteral`?

math-fehr replied:
Oh yes, that makes more sense, I'll fix that!

math-fehr committed:
Make name a parameter of Builtin_Attr

joker-eph commented on mlir/lib/IR/MLIRContext.cpp:
```diff
@@ -212,6 +212,9 @@ class MLIRContextImpl {
   StorageUniquer typeUniquer;
 
+  /// This is a mapping from type name to the abstract type describing it.
+  llvm::StringMap<AbstractType *> nameToType;
```
Can we use `DenseMap<StringAttr, AbstractType *> nameToType;` (more efficient storage, and more efficient lookup)?

Also when is it useful? (please add to the comment: the "what" is less important than the "why").

math-fehr replied:
Added!

joker-eph replied:
did you?

math-fehr replied:
Oh my bad, it got erased again in a following PR, I'll add it back now!
Sorry about that!

math-fehr committed:
Use StringLiteral instead of StringRef, and use fields rather than methods

math-fehr commented:
I moved `getTypeName` and `getAttrName` to just a `name` field, so that it will be the same for both attributes and types, and also would not clash with any `name` parameter that could be define in ODS.
Do you think that make sense, or do you think that this is too much different than operations (that have a `getName` static method). Or do you think this should also be changed in operations?

Also, I added the change in `MLIRContext` so that we store a `DensMap` of `StringAttr` rather than a `StringMap`.

Currently, I have a `Address boundary error` when `build/NATIVE/bin/mlir-linalg-ods-yaml-gen` is being used in the build. I'm trying to figure out if it's because of the `::name` change, or the `StringAttr`.

math-fehr committed:
Remove use of StringAttr in the MLIRContext and fix one attr

math-fehr commented:
So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.

For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.

So for now I removed it, please tell me if you see a way I could make this work!

math-fehr committed:
Add the missing AbstractAttribute::lookup

math-fehr committed:
Add a test for AbstractType::lookup and AbstractAttribute::lookup.

Mogball commented on mlir/unittests/IR/TypeAttrNamesTest.cpp:
```diff
@@ -0,0 +1,90 @@
+  // Check that lookups of non-existing Attributes returns nullopt.
+  // Even if an attribute with the same name exists.
+  EXPECT_FALSE(AbstractAttribute::lookup("fake.foo", &ctx).has_value());
+}
```
newline

math-fehr committed:
Add missing newline

math-fehr commented:
@joker-eph @kuhar, does this PR sounds good to you now?
I'll wait for your approval to merge it, but otherwise I think enough time passed that anyone interested commented?

joker-eph commented:
> So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.
> 
> For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.

Oh great... We can't add the name for `StringAttr` before registering the storage is that it?

Pretty sad about these string duplication, but we can live with it. Please document this next to the StringMap.

> So for now I removed it, please tell me if you see a way I could make this work!

It's likely possible, but will require some deeper look. I may try this after you land it!

math-fehr committed:
Add documentation that got removed

math-fehr commented:
> > So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.
> > For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.
> 
> Oh great... We can't add the name for `StringAttr` before registering the storage is that it?
> 
> Pretty sad about these string duplication, but we can live with it. Please document this next to the StringMap.
> 
> > So for now I removed it, please tell me if you see a way I could make this work!
> 
> It's likely possible, but will require some deeper look. I may try this after you land it!

Another solution would be to use a `DenseMap` of `StringRef` instead, as all names are basically `StringLiteral` for C++ defined attributes, and `StringRef` for dynamically defined attributes (as they already store their names).
Do you think that makes more sense for this PR?

math-fehr committed:
Comment why we don't use a DenseMap of StringAttr

kuhar reviewed (approved):
LGTM

joker-eph commented:
> Do you think that makes more sense for this PR?

We could if we're confident this is enough. The API should then take a "StringLiteral" (it's more specific than a StringRef) I believe.

math-fehr commented:
> > Do you think that makes more sense for this PR?
> 
> We could if we're confident this is enough. The API should then take a "StringLiteral" (it's more specific than a StringRef) I believe.

I'm confident `StringRef` will work, since it works for both statically-defined and dynamically defined types and attributes.
For `StringLiteral`, it wouldn't work for dynamically defined attributes and types.

math-fehr committed:
Do not duplicate the names of types and attributes

joker-eph commented:
Well it's a low-level API, so let's document the lifetime constraint on the StringRef then!

math-fehr committed:
Document lifetimes of StringRef

makslevental commented:
Congrats @math-fehr on landing a solid patch. I'll be using/abusing this soon for the Python bindings :)
